eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/NoImplInPublicAPI.java

```diff
@@ -0,0 +1,114 @@
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class NoImplInPublicAPI extends AbstractCheck {
+
+    private static final String COM_AZURE = "com.azure";
+    private static final String DOT_IMPLEMENTATION = ".implementation";
+    private static final String IMPL_PACKAGE_REGULAR_EXPR = "^com\\.azure(\\..*)+\\.implementation(\\..*)*";
+    private static final String PARAM_TYPE_ERROR = "\"%s\" class is in an implementation package, and it should not be used as a parameter type in public API. Alternatively, it can be removed from the implementation package and made public API, after appropriate API review.";
+    private static final String RETURN_TYPE_ERROR = "\"%s\" class is in an implementation package, and it should not be a return type from public API. Alternatively, it can be removed from the implementation package and made public API.";
+
+    private static boolean isTrackTwo;
+    private static boolean isImplPackage;
+    private Set<String> implementationClassSet = new HashSet<>();
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {
+            TokenTypes.PACKAGE_DEF,
+            TokenTypes.IMPORT,
+            TokenTypes.METHOD_DEF
+        };
+    }
+
+    @Override
+    public void beginTree(DetailAST root) {
+        this.isImplPackage = false;
+        this.isTrackTwo = false;
+        this.implementationClassSet.clear();
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        if (ast.getType() == TokenTypes.PACKAGE_DEF) {
+            String packageName = FullIdent.createFullIdent(ast.findFirstToken(TokenTypes.DOT)).getText();
+            this.isTrackTwo = packageName.startsWith(COM_AZURE);
+            this.isImplPackage = packageName.contains(DOT_IMPLEMENTATION);
+            return;
+        } else {
+            if (this.isTrackTwo) {
+                if (this.isImplPackage) {
+                    return;
+                }
+            } else {
+                return;
+            }
+        }
+
+        switch (ast.getType()) {
+            case TokenTypes.IMPORT:
+                String importClassPath = FullIdent.createFullIdentBelow(ast).getText();
+                if (importClassPath.startsWith(COM_AZURE)) {
+                    int idx = importClassPath.indexOf(DOT_IMPLEMENTATION);
+                    if (idx > -1) {
+                        String remainingPath = importClassPath.substring(idx + DOT_IMPLEMENTATION.length());
+                        if (remainingPath != null && remainingPath.length() > 1) {
+                            String className = remainingPath.substring(remainingPath.lastIndexOf(".") + 1);
+                            implementationClassSet.add(className);
```

Comment:
I don't like this approach of looking through the import list. It means you might miss fully-qualified class references to implementation classes. I would much rather you inspect the type in the `METHOD_DEF` branch down directly and not have the `implementationClassSet` collection at all.

Reply:
If you are unable to find a way to do this, just leave the code as is and we can return to it later